api/src/main/java/org/pmiops/workbench/api/CloudTaskFreeCreditExpiry.java

```diff
@@ -0,0 +1,307 @@
+    Set<DbUser> freeTierUsers = getFreeTierActiveWorkspaceCreatorsIn(allUsers);
+
+    // Find users who exceeded their free tier limit
+    // Here costs in liveCostByCreator could be outdated because we're filtering on active  or
+    // recently deleted workspaces in previous steps.
+    // However, dbCostByCreator will contain the up-to-date costs for all the
+    // other workspaces. This is why Math.max is used
+    final Set<DbUser> expiredUsers =
+        dbUsersWithChangedCosts.entrySet().stream()
+            .filter(
+                e ->
+                    CostComparisonUtils.costAboveLimit(
+                        e.getValue(),
+                        Math.max(
+                            dbCostByCreator.get(e.getKey()), liveCostByCreator.get(e.getKey())),
+                        workbenchConfig.get().billing.defaultFreeCreditsDollarLimit))
+            .map(Map.Entry::getValue)
+            .collect(Collectors.toSet());
+
+    final Set<DbUser> newlyExpiredFreeTierUsers = Sets.intersection(expiredUsers, freeTierUsers);
+
+    logger.info(
+        String.format(
+            "Found %d users exceeding their free tier limit, out of which, %d are new",
+            expiredUsers.size(), newlyExpiredFreeTierUsers.size()));
+
+    return newlyExpiredFreeTierUsers;
+  }
+
+  /**
+   * Find the users with db costs different from the live costs
+   *
+   * @param dbCostByCreator Map of userId->dbCost
+   * @param liveCostByCreator Map of userId->liveCost
+   * @return a {@link Map} of user Ids to their DbUsers
+   */
+  @NotNull
+  private Map<Long, DbUser> findDbUsersWithChangedCosts(
+      Map<Long, Double> dbCostByCreator, Map<Long, Double> liveCostByCreator) {
+    final Set<Long> usersWithChangedCosts =
+        liveCostByCreator.keySet().stream()
+            .filter(
+                u ->
+                    CostComparisonUtils.compareCosts(
+                            dbCostByCreator.get(u), liveCostByCreator.get(u))
+                        != 0)
+            .collect(Collectors.toSet());
+
+    logger.info(String.format("Found %d users with changed costs.", usersWithChangedCosts.size()));
+
+    Map<Long, DbUser> dbUsersWithChangedCosts =
+        StreamSupport.stream(userDao.findAllById(usersWithChangedCosts).spliterator(), false)
+            .collect(Collectors.toMap(DbUser::getUserId, Function.identity()));
+    return dbUsersWithChangedCosts;
+  }
+
+  private Set<DbUser> getFreeTierActiveWorkspaceCreatorsIn(Set<DbUser> users) {
+    return workspaceDao.findCreatorsByBillingStatusAndBillingAccountNameIn(
+        BillingStatus.ACTIVE,
+        new ArrayList<>(workbenchConfig.get().billing.freeTierBillingAccountNames()),
+        users);
+  }
+
+  private void deleteAppsAndRuntimesInFreeTierWorkspaces(DbUser user) {
+    logger.info("Deleting apps and runtimes for user " + user.getUsername());
+
+    impersonatedWorkspaceService.getOwnedWorkspaces(user.getUsername()).stream()
```

Comment:
I don't think this is the right function to use. My understanding is that function will give you the workspaces you are the **owner** of. But in our case, we want **creator**. 
We can just get this from our DB.
The above method might be the right one. `workspaceDao.findAllByCreator`